src/lib/Microsoft.Fx.Portability.Reports.DGML/ReferenceNode.cs

```diff
@@ -66,65 +66,65 @@ public double GetPortabilityIndexForReferences(int target)
-            {
-                // Cycle detected
-                return 1;
-            }
-            else
-            {
-                // remove the calls from the current node.
-                availableApis -= UsageData[target].GetAvailableAPICalls();
-                unavailableApis -= UsageData[target].GetUnavailableAPICalls();
+            GetAPICountFromReferences(target, out int availableApis, out int unavailableApis);
 
-                // prevent Div/0
-                if (availableApis == 0 && unavailableApis == 0)
-                    return 0;
+            // prevent Div/0. When availableApis is 0 and unavailableApis is 0, it likely means that it's references are all not resulted assemblies, return 0.
+            if (availableApis == 0 && unavailableApis == 0)
+                return 0;
 
-                return availableApis / ((double)availableApis + unavailableApis);
-            }
+            return availableApis / ((double)availableApis + unavailableApis);
         }
 
-        public bool TryGetAPICountFromReferences(int target, out int availAPIs, out int unavailAPIs)
+        private void GetAPICountFromReferences(int target, out int availAPIs, out int unavailAPIs)
         {
-            availAPIs = UsageData[target].GetAvailableAPICalls();
-            unavailAPIs = UsageData[target].GetUnavailableAPICalls();
-
-            // We are going to use a flag on the object to detect if we have a reference cycle while computing the APIs for the references.
-            if (_searchInGraph == true)
-            {
-                // Cycle!!!
-                _searchInGraph = false; // Reset this flag
-                return false;
-            }
-            else
-            {
-                _searchInGraph = true;
-            }
+            availAPIs = 0;
+            unavailAPIs = 0;
 
             foreach (var item in Nodes)
             {
-                if (item.UsageData == null)
+                // We are going to use the flag on the object to detect if we have run into this node in the reference graph while computing the APIs for the references,
+                // because we want to only count a reference node once if it is referenced in the graph more than once.
+                if (item._searchInGraph == true)
                 {
-                    // skip the Node with no UsageData
                     continue;
                 }
-
-                if (!item.TryGetAPICountFromReferences(target, out int refCountAvail, out int refCountUnavail))
+                else
                 {
-                    // Cycle!
-                    _searchInGraph = false; // Reset this flag
+                    item._searchInGraph = true;
+                }
 
-                    return false;
+                if (item.UsageData != null)
+                {
+                    availAPIs += item.UsageData[target].GetAvailableAPICalls();
```

Comment:
Can any of the rest of these calls fail? (can [target] fail with index out of range or another exception, or return null?

Reply:
Fixed in the new commit.

Reply:
Did this actually get fixed? Or we just proved these calls would always succeed?

Reply:
Fixed by adding try-catch around item.UsageData[target] statements in the last commit. Any additional concern?